lang/cpp.go

```diff
@@ -0,0 +1,374 @@
+	return strings.Count(t, "[]"), cppTypes[strings.ReplaceAll(t, "[]", "")]
+}
+
+func (c cpp) getVectorTypeName(d int, t string) string {
+	return strings.Repeat("vector<", d) + t + strings.Repeat(">", d)
+}
+
+func (c cpp) getDeclCodeForType(d int, t string, n string) string {
+	return fmt.Sprintf("%s %s;", c.getVectorTypeName(d, t), n)
+}
+
+func (c cpp) getScanCodeForType(d int, t string, n string, ifs string) string {
+	if d == 0 && t == "string" {
+		return fmt.Sprintf("%s >> quoted(%s);", ifs, n)
+	} else {
+		return fmt.Sprintf("%s >> %s;", ifs, n)
+	}
+}
+
+func (c cpp) getPrintCodeForType(d int, t string, n string, ofs string) string {
+	if d == 0 {
+		if t == "string" {
+			return fmt.Sprintf("%s << quoted(%s);", ofs, n)
+		}
+		if t == "double" {
+			return fmt.Sprintf("{ char buf[320]; sprintf(buf, \"%%.5f\", %s); %s << string(buf); }", n, ofs)
+		}
+		return fmt.Sprintf("%s << %s;", ofs, n)
+	} else {
+		return fmt.Sprintf("%s << %s;", ofs, n)
+	}
+}
+
+func (c cpp) getParamString(params []leetcode.MetaDataParam) string {
+	var paramList []string
+	for _, param := range params {
+		paramList = append(paramList, param.Name)
+	}
+	return strings.Join(paramList, ", ")
+}
+
+func (c cpp) generateScanCode(q *leetcode.QuestionData) string {
+	if q.MetaData.SystemDesign {
+		return fmt.Sprintf(
+			"\t%s %s\n",
+			c.getDeclCodeForType(1, "string", systemDesignFuncNames),
+			c.getScanCodeForType(1, "string", systemDesignFuncNames, inputStreamName),
+		)
+	}
+
+	var scanCode string
+	for _, param := range q.MetaData.Params {
+		dimCnt, cppType := c.getCppTypeName(param.Type)
+		scanCode += fmt.Sprintf(
+			"\t%s %s\n",
+			c.getDeclCodeForType(dimCnt, cppType, param.Name),
+			c.getScanCodeForType(dimCnt, cppType, param.Name, inputStreamName),
+		)
+	}
+	return scanCode
+}
+
+func (c cpp) generateInitCode(q *leetcode.QuestionData) string {
+	if q.MetaData.SystemDesign {
+		return fmt.Sprintf("\t%s *%s;\n", q.MetaData.ClassName, objectName)
+	} else {
+		return fmt.Sprintf("\tSolution *%s = new Solution();\n", objectName)
+	}
+}
+
+func (c cpp) generateCallCode(q *leetcode.QuestionData) string {
+	var callCode string
+
+	generateParamScanningCode := func(params []leetcode.MetaDataParam) {
+		if len(params) > 0 {
+			for _, param := range params {
+				dimCnt, cppType := c.getCppTypeName(param.Type)
+				callCode += fmt.Sprintf(
+					"\t\t\t%s %s %s.ignore();\n",
+					c.getDeclCodeForType(dimCnt, cppType, param.Name),
+					c.getScanCodeForType(dimCnt, cppType, param.Name, inputStreamName),
+					inputStreamName,
+				)
+			}
+		} else {
+			callCode += fmt.Sprintf("\t\t\t%s.ignore();\n", inputStreamName)
+		}
+	}
+
+	if q.MetaData.SystemDesign {
+		className := q.MetaData.ClassName
+		callCode += fmt.Sprintf("\tauto hash_value_%s = hash<string>()(\"%s\");\n", className, className)
+		for _, method := range q.MetaData.Methods {
+			callCode += fmt.Sprintf("\tauto hash_value_%s = hash<string>()(\"%s\");\n", method.Name, method.Name)
```

Comment:
Can we make the code compare strings directly (is it possible is C++?) instead of using the hash-then-compare approach? The hash method makes the code lengthy and unclear. And regarding the performance concerns, local testing isn't meant to perfect since it only has a limited number of cases and short inputs, so I don't think it would make a significant difference.

Reply:
I agree, I will change this in the next commit.
This has been an issue of C++ since the begining, because its switch statements are derieved from C, which only supports integer types. And there's no compile-time hashing in C++ standard library.

Reply:
What do you think of this approach? This makes the code shorter than `if else` while maintaining performance and readability:
```cpp

// define methods
const unordered_map<string, function<void()>> methods = { 
{ "LFUCache", [&]() {
	int capacity; cin >> capacity; cin.ignore();
	obj = new LFUCache(capacity);
	out_stream << "null,";
} },
{ "get", [&]() {
	int key; cin >> key; cin.ignore();
	out_stream << obj->get(key); out_stream << ',';
} },
{ "put", [&]() {
	int key; cin >> key; cin.ignore();
	int value; cin >> value; cin.ignore();
	obj->put(key, value);
	out_stream << "null,";
} },
};

// invoke methods
cin.ignore(); out_stream << '[';
for (auto &&sys_design_func : sys_design_funcs) {
cin.ignore();
methods.at(sys_design_func)();
cin.ignore();
}
out_stream.seekp(-1, ios_base::end); out_stream << ']';
```

Reply:
looks great, it's really nice